airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java

```diff
@@ -189,13 +206,13 @@ private boolean hasSnapshotFinished(final JsonNode eventAsJson) {
     return !SnapshotMetadata.isSnapshotEventMetadata(snapshotMetadata);
   }
 
-  private void requestClose() {
-    try {
-      requestClose.call();
-    } catch (final Exception e) {
-      throw new RuntimeException(e);
+  private void requestClose(final String closeLogMessage) {
+    if (signalledDebeziumEngineShutdown) {
+      return;
     }
-    throwExceptionIfSnapshotNotFinished();
+    LOGGER.info(closeLogMessage);
+    debeziumShutdownProcedure.initiateShutdownProcedure();
+    signalledDebeziumEngineShutdown = true;
   }
```

Comment:
From my point of view we should be aware that `requestClose` is called only once in the code and not just figure out if it has been called before using an if statement and the flag `signalledDebeziumEngineShutdown`.

Reply:
We call `requestClose` when certain conditions are met. For instance if the LSN of a record is higher than the target LSN that we  aimed for. Please remember that Debezium doesn't understand the concept of target LSN and it will continue to fetch data until it's shutdown. Also Debezium is running in a separate thread. So often Debezium pulls in multiple records which are after the target LSN and only when we see the first of the record which is after target LSN, we ask Debezium to shutdown but by that time Debezium has already fetched even more records. And we need to process these records. But each of the record that we process from that point is after the target LSN which makes the condition to call `requestClose` true. Here is sample screenshot of that happening. 
<img width="1333" alt="Screenshot 2023-03-23 at 12 49 39 PM" src="https://user-images.githubusercontent.com/28527634/227131969-bc937dc1-3e1b-45f5-9d51-ccbee10d36e6.png">

As you can see in the above screenshot we processed three such records which had been pulled which were after the target LSN and `requestClose` was called thrice. To tackle this, we need the boolean flag.